### PR TITLE
tests: run most smoke tests on devtools

### PR DESCRIPTION
**Pull request by connorjclark**
Just `seo` for now.

Expanded  `yarn run-devtools`:
- Saves artifacts
- Inject custom config
- Some extra waiting to fix a bug where multiple launches of `yarn run-devtools` resulted in the script breaking

ref #13539
ref #13396

### CODE REVIEW

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+const execFilePromise = util.promisify(execFile);
+
+/** @type {Promise<void>} */
+let setupPromise;
+async function setup() {
+  if (setupPromise) return setupPromise;
+
+  setupPromise = new Promise(resolve => {
+    // execFileSync('yarn', ['open-devtools']);
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:25]: TODO: I thought about doing the setup for building ToT devtools here... would that be useful?

File: lighthouse-core/scripts/open-devtools.sh [hunk 2]
```diff
@@ -8,6 +8,14 @@ set -euo pipefail
...
 ##
 
+# 1) Builds Lighthouse for DevTools
+# 2) Rolls to local devtools repo. By default, this is the temporary checkout in .tmp
+# 3) Builds devtools frontend with new Lighthouse roll
+# 4) Opens $CHROME_PATH using new devtools frontend build, passing any additional args to Chrome.
+# 
+# Run `yarn test-devtools` first to update the temporary devtools checkout.
+# Specify `$DEVTOOLS_PATH` to use a different devtools repo.
+
```
Comment (connorjclark) [inline lighthouse-core/scripts/open-devtools.sh:18]: some driveby extra documentation

File: lighthouse-core/scripts/pptr-run-devtools.js [hunk 3]
```diff
@@ -147,6 +160,33 @@ async function testPage(page, browser, url) {
       .catch(reject);
   });
 
+  if (config) {
```
Comment (adamraine) [inline lighthouse-core/scripts/pptr-run-devtools.js:163]: Could you add a comment about why we need to attach to a different target?

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:25]: > TODO: I thought about doing the setup for building ToT devtools here... would that be useful?

yeah, though it would be nice if it didn't attempt to re-update on every run. Is it currently only the content-shell we only update every X commits?

File: lighthouse-cli/test/smokehouse/frontends/smokehouse-bin.js [hunk 4]
```diff
@@ -36,6 +36,7 @@ const coreTestDefnsPath =
 const runnerPaths = {
   cli: '../lighthouse-runners/cli.js',
   bundle: '../lighthouse-runners/bundle.js',
+  devtools: '../lighthouse-runners/devtools.js',
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/frontends/smokehouse-bin.js:39]: add to the smokehouse readme ascii art? :)

**Comment by brendankenny** [comment]
this is great!

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  const args = [
+    'run-devtools',
+    url,
+    '--output-dir', outputDir,
+  ];
+  if (configJson) {
+    args.push('--config');
+    args.push(JSON.stringify(configJson));
+  }
+  await execFilePromise('yarn', args);
```
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:50]: no, we just rely on lighthouse's timeout, which wouldn't save us here

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ * @return {Promise<{lhr: LH.Result, artifacts: LH.Artifacts, log: string}>}
+ */
+async function runLighthouse(url, configJson, testRunnerOptions = {}) {
+  await setup();
+
+  const outputDir = fs.mkdtempSync(os.tmpdir() + '/lh-smoke-cdt-runner-');
+  const args = [
+    'run-devtools',
+    url,
+    '--output-dir', outputDir,
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:44]: Ah, I keep making this mistake, thinking that is the default behavior. It is for  open-devtools! I want to followup with a change that does that.

Will add the arg here for now.

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:50]: I'll add a timeout here!

File: .github/workflows/devtools.yml [hunk 7]
```diff
@@ -73,3 +73,18 @@ jobs:
       with:
         name: results
         path: ${{ github.workspace }}/lighthouse/.tmp/layout-test-results
+    
+    # Run smoke tests via DevTools
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply .github/workflows/devtools.yml:77]: How to not avoid redoing all the work for building devtools?

File: .github/workflows/devtools.yml [hunk 8]
```diff
@@ -73,3 +73,18 @@ jobs:
...
+    - name: Install Chrome ToT
+      working-directory: /home/runner
+      run: bash $GITHUB_WORKSPACE/lighthouse-core/scripts/download-chrome.sh && mv chrome-linux chrome-linux-tot
+    - run: sudo apt-get install xvfb
+    - name: yarn smoke --runner devtools
+      # TODO: run on all tests.
+      # - Current DevTools hangs on any page with a service worker.
+      #   https://github.com/GoogleChrome/lighthouse/issues/13396
+      # - Should undock devtools because many expectations are sensitive to the default viewport size.
+      run: xvfb-run --auto-servernum yarn smoke --runner devtools seo
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply .github/workflows/devtools.yml:90]: dbw needs the undock thing, seo was pretty non problematic

[the remark this answers is not in the document]
Comment (connorjclark) [reply .github/workflows/devtools.yml:90]: I'll add the timeout + the undock code and run everything, then disable the ones that fail atm

File: lighthouse-core/scripts/pptr-run-devtools.js [hunk 9]
```diff
@@ -105,6 +112,13 @@ const startLighthouse = `
   const panel = UI.panels.lighthouse || UI.panels.audits;
   const button = panel.contentElement.querySelector('button');
   if (button.disabled) throw new Error('Start button disabled');
+
+  UI.dockController.setDockSide('undocked');
+
+  // Give the main target model a moment to be available.
+  // Otherwise, 'SDK.TargetManager.TargetManager.instance().mainTarget()' is null.
+  await new Promise(resolve => setTimeout(resolve, 1000));
```
Comment (adamraine) [inline lighthouse-core/scripts/pptr-run-devtools.js:120]: Could we create a [target observer](https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/core/sdk/TargetManager.ts;l=155-163?q=targetmanager&ss=chromium%2Fchromium%2Fsrc:third_party%2Fdevtools-frontend%2Fsrc%2F) for this instead? I'd like to avoid waiting for arbitrary amounts of time if possible.

File: lighthouse-core/scripts/pptr-run-devtools.js [hunk 10]
```diff
@@ -117,7 +117,17 @@ const startLighthouse = `
 
   // Give the main target model a moment to be available.
   // Otherwise, 'SDK.TargetManager.TargetManager.instance().mainTarget()' is null.
-  await new Promise(resolve => setTimeout(resolve, 1000));
+  const targetManager =
+    SDK.targetManager || (SDK.TargetManager.TargetManager || SDK.TargetManager).instance();
+  if (targetManager.mainTarget() === null) {
+    if (targetManager.observeTargets) {
```
Comment (adamraine) [inline lighthouse-core/scripts/pptr-run-devtools.js:123]: When would this be falsy? In an older version of DT?

File: lighthouse-core/scripts/pptr-run-devtools.js [hunk 11]
```diff
@@ -117,7 +117,17 @@ const startLighthouse = `
...
+      await new Promise(resolve => targetManager.observeTargets({targetAdded: resolve}));
```
Comment (adamraine) [inline lighthouse-core/scripts/pptr-run-devtools.js:124]: Just to be on the safe side, could you add a blank `targetRemoved` listener

File: lighthouse-core/scripts/pptr-run-devtools.js (hunk 10 above)
Comment (connorjclark) [reply lighthouse-core/scripts/pptr-run-devtools.js:123]: Could be. It was pretty involved making sure a bunch of older chromes worked with this script, so I am doing without testing if it is necessary 😬

File: lighthouse-core/scripts/pptr-run-devtools.js (hunk 11 above)
Comment (connorjclark) [reply lighthouse-core/scripts/pptr-run-devtools.js:124]: would it error without?

Comment (adamraine) [reply lighthouse-core/scripts/pptr-run-devtools.js:124]: `targetRemoved` is expected here, would throw when it tries to run an unset property:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/core/sdk/TargetManager.ts;l=215?q=TargetManager&ss=chromium%2Fchromium%2Fsrc:third_party%2Fdevtools-frontend%2Fsrc%2F

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:24]: The GHA workflow already builds devtools before starting this smoke test as part of the layout tests ... so let's do something here but only for local runs.

```
export DEVTOOLS_PATH=".tmp/devtools/devtools-frontend"
node lighthouse-core/test/chromium-web-tests/download-devtools.sh
node lighthouse-core/test/chromium-web-tests/roll-devtools.sh
```

(and then the local devtools smoke run would defer to w/e CHROME_PATH you have, not use the latest content shell ... unless thats your chrome path.)

File: .github/workflows/devtools.yml [hunk 12]
```diff
@@ -73,3 +73,19 @@ jobs:
...
+    - name: Install Chrome ToT
+      working-directory: /Users/runner
+      run: bash $GITHUB_WORKSPACE/lighthouse/lighthouse-core/scripts/download-chrome.sh && mv chrome-mac chrome-mac-tot
+    - run: mkdir latest-run
+      working-directory: ${{ github.workspace }}/lighthouse
+    - name: yarn smoke --runner devtools
+      # TODO: run on all tests.
+      # - Current DevTools hangs on any page with a service worker.
+      #   https://github.com/GoogleChrome/lighthouse/issues/13396
+      # - Various other issues that needed investigation.
```
Comment (adamraine) [inline .github/workflows/devtools.yml:89]: Can you create an issue to track this second bullet?

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * Launch Chrome and do a full Lighthouse run via DevTools.
+ * @param {string} url
+ * @param {LH.Config.Json=} configJson
+ * @param {{isDebug?: boolean}=} testRunnerOptions
+ * @return {Promise<{lhr: LH.Result, artifacts: LH.Artifacts, log: string}>}
+ */
+async function runLighthouse(url, configJson, testRunnerOptions = {}) {
+  const outputDir = fs.mkdtempSync(os.tmpdir() + '/lh-smoke-cdt-runner-');
```
Comment (adamraine) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:64]: Does this create multiple temp directories if the smokes are running concurrently?

**Review by adamraine — APPROVED** [review]
LGTM

File: .github/workflows/devtools.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply .github/workflows/devtools.yml:77]: > How to avoid redoing all the work for building devtools?

Yeah it's not pretty. We could cache the build (or have dependent jobs)? I'm not sure of the best approach, but if the devtools CI job is normally going to take 30+ minutes we'll definitely want to split it up.

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js (hunk 13 above)
Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:64]: yeah, [fs.mkdtempSync`](https://nodejs.org/api/fs.html#fsmkdtempprefix-options-callback) handles that

File: lighthouse-cli/test/smokehouse/smokehouse.js [hunk 14]
```diff
@@ -146,8 +151,19 @@ async function runSmokeTest(smokeTestDefn, testOptions) {
 
     // Run Lighthouse.
     try {
+      /** @type {NodeJS.Timeout} */
+      let timeout;
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/smokehouse.js:155]: I mentioned this out of band, but for tracking here:

This adds some timeout redundancy here, since Lighthouse itself will also fail with an error after it times out, which leaves it somewhat unclear where modified timeouts should be specified, here or in the runners. It's not a huge deal, (100s is pretty long, so the core Lighthouse one will always end first), but simplicity-wise it might make sense to move this timeout into the new `devtools.js` and leave runs with the other runners unchanged.

This was mostly a nit before, but the introduction of `beforeAll` makes it a bigger win simplicity-wise, since `devtools.js` wouldn't need a `beforeAll` at that point, it could just call those lines itself and just exempt them from the overall timeout.

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+  /** @type {Promise<void>} */
+  const promise = new Promise((resolve, reject) => {
+    const spawnHandle = spawn(command, args);
+    spawnHandle.on('close', code => {
+      if (code === 0) resolve();
+      else reject(new Error(`Command exited with code ${code}`));
+    });
+    spawnHandle.on('error', reject);
+    spawnHandle.stdout.on('data', data => {
+      console.log(data.toString());
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:34]: are these lines intended to stay or for debugging?

File: lighthouse-core/scripts/pptr-run-devtools.js [hunk 16]
```diff
@@ -228,6 +295,12 @@ async function run() {
   if (customDevtools) {
     console.log(`Using custom devtools frontend: ${customDevtools}`);
     console.log('Make sure it has been built recently!');
+    if (!customDevtools.startsWith('file://')) {
+      throw new Error('custom-devtools-frontend must be a file:// URL');
+    }
+    if (!fs.existsSync(fileURLToPath(customDevtools))) {
```
Comment (adamraine) [inline lighthouse-core/scripts/pptr-run-devtools.js:301]: Running locally on mac I get an error here:

```
(node:43463) UnhandledPromiseRejectionWarning: TypeError [ERR_INVALID_FILE_URL_HOST]: File URL host must be "localhost" or empty on darwin
    at new NodeError (internal/errors.js:322:7)
    at getPathFromURLPosix (internal/url.js:1350:11)
    at fileURLToPath (internal/url.js:1373:50)
    at run (file:///Users/asraine/lighthouse/lighthouse-core/scripts/pptr-run-devtools.js:301:24)
    at file:///Users/asraine/lighthouse/lighthouse-core/scripts/pptr-run-devtools.js:353:1
    at ModuleJob.run (internal/modules/esm/module_job.js:183:25)
    at async Loader.import (internal/modules/esm/loader.js:178:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5)
```
Am I missing something, tried on Node 14 & 16.

Comment (connorjclark) [reply lighthouse-core/scripts/pptr-run-devtools.js:301]: Works for me.

```
yarn run-devtools --custom-devtools-frontend=file://$PWD/.tmp/chromium-web-tests/devtools/devtools-frontend/out/Default/gen/front_end https://www.example.com
```

node 14.8.1

Comment (adamraine) [reply lighthouse-core/scripts/pptr-run-devtools.js:301]: I think it's because a relative path is set in `DEVTOOLS_PATH` if run locally.

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js (hunk 15 above)
Comment (connorjclark) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:34]: Yes, otherwise the cases where things hang or error aren't printed (the runner interface only allows returning successful logs). and it's nicer to see the output stream in.

File: lighthouse-core/scripts/pptr-run-devtools.js (hunk 16 above)
Comment (connorjclark) [reply lighthouse-core/scripts/pptr-run-devtools.js:301]: What is your command?

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js (hunk 15 above)
Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:34]: just checking that it's intentional. The tradeoff is that concurrent runs will be streaming in stuff all interleaved with each other, but that can be helpful, especially when things are new and need to check that it's actually running, etc

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+ * Launch Chrome and do a full Lighthouse run via DevTools.
+ * @param {string} url
+ * @param {LH.Config.Json=} configJson
+ * @param {{isDebug?: boolean}=} testRunnerOptions
+ * @return {Promise<{lhr: LH.Result, artifacts: LH.Artifacts, log: string}>}
+ */
+async function runLighthouse(url, configJson, testRunnerOptions = {}) {
+  const outputDir = fs.mkdtempSync(os.tmpdir() + '/lh-smoke-cdt-runner-');
+  const devtoolsDir =
+    process.env.DEVTOOLS_PATH || `${LH_ROOT}/.tmp/chromium-web-tests/devtools/devtools-frontend`;
```
Comment (adamraine) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:66]: Using `path.resolve` here would fix my other comment

File: lighthouse-core/scripts/pptr-run-devtools.js (hunk 16 above)
Comment (adamraine) [reply lighthouse-core/scripts/pptr-run-devtools.js:301]: `yarn smoke --runner devtools csp`

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    });
+  });
+  await promise;
+
+  return log;
+}
+
+/** @type {Promise<void>} */
+let beforeAllPromise;
+async function beforeAll() {
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:52]: can you add a jsdoc like, "Build the lighthouse DevTools bundle and fetch and build the DevTools frontend using it" or whatever?

File: lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+  if (process.env.CI) return;
+
+  process.env.DEVTOOLS_PATH = devtoolsDir;
+  await spawnAndLog('bash', ['lighthouse-core/test/chromium-web-tests/download-devtools.sh']);
+  await spawnAndLog('bash', ['lighthouse-core/test/chromium-web-tests/roll-devtools.sh']);
+}
+
+/**
+ * Launch Chrome and do a full Lighthouse run via DevTools.
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/devtools.js:61]: I'm having trouble coming up with a non-awkward way to say this, but seems worth clarifying it's the current/local Lighthouse devtools bundle run in the latest DevTools frontend
vs
latest DevTools frontend with whatever Lighthouse version was last rolled into Chromium.

File: lighthouse-core/scripts/pptr-run-devtools.js [hunk 20]
```diff
@@ -113,6 +121,32 @@ const startLighthouse = `
...
   if (button.disabled) throw new Error('Start button disabled');
+
+  UI.dockController.setDockSide('undocked');
+
+  // Give the main target model a moment to be available.
+  // Otherwise, 'SDK.TargetManager.TargetManager.instance().mainTarget()' is null.
+  if (self.runtime && self.runtime.loadLegacyModule) {
+    // This exposes TargetManager via self.SDK.
+    try {
+      await self.runtime.loadLegacyModule('core/sdk/sdk-legacy.js');
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/scripts/pptr-run-devtools.js:132]: Sure hope not, otherwise I found no way to access TargetManager!
